### PR TITLE
Consistent naming, minor additions

### PR DESCRIPTION
**Pull request by riedel-ferringer**
suggestion to make naming more clear and consistent: always use actual/expected instead of L/R
minor hopefully useful addions to the ScoreBoard and TbUtilPkg

### CODE REVIEW

File: TbUtilPkg.vhd
Comment (Paebbels) [inline TbUtilPkg.vhd:521]: `AckTime` is unused, right?

File: TbUtilPkg.vhd
Comment (Paebbels) [inline TbUtilPkg.vhd:637]: `AckTime` is unused, right?

**Review by Paebbels — CHANGES_REQUESTED** [review]

**Comment by Paebbels** [comment]
Thanks for you proposals.

@JimLewis please have a look.

File: TbUtilPkg.vhd
Comment (riedel-ferringer) [reply TbUtilPkg.vhd]: Right. Corrected.

File: TbUtilPkg.vhd
Comment (riedel-ferringer) [reply TbUtilPkg.vhd]: Right. Corrected.

**Review by JimLewis — CHANGES_REQUESTED** [review]

File: ScoreboardGenericPkg.vhd
Comment (JimLewis) [inline ScoreboardGenericPkg.vhd:2086]: Purpose?  Use model?

Not for or against, just want to understand.    Also note that the internal Error and Check counts are there because the implementation pre-dated the AlertLogPkg.   I suspect that a future implementation of the ScoreboardPkg that is just a singleton implementation will use only the AlertLogPkg capability, so we need to understand why here so we understand the design considerations for the singleton.  

You only added it to the protected type API.   I don't see anything that adds it to the Singleton API.

**Review by JimLewis — CHANGES_REQUESTED** [review]

File: TbUtilPkg.vhd
Comment (JimLewis) [inline TbUtilPkg.vhd:645]: Did the AckTime variable get deleted (from here and the original std_logic one)?    Since it is unused, it should be removed.

File: TbUtilPkg.vhd
Comment (JimLewis) [inline TbUtilPkg.vhd]: Shouldn't the code show AckTime as removed - both variable and the assignment?   I am flagging it only because in the review doc it is shown as there.

**Review by JimLewis — COMMENTED** [review]
I am still seeing AckTime in the new WaitForTransaction Procedures.   Is it really there or is it just a feature of GitHub?

File: AlertLogPkg.vhd
Comment (JimLewis) [inline AlertLogPkg.vhd:3822]: This illuminates an inconsistency between AlertIfEqual and AffirmIfEqual.   AffirmIfEqual uses to_hstring for arrays of std_logic, however, AlertIfEqual uses to_string.  

Would you be willing to make this update.   

to_integer is an interesting option - until the array size gets larger than 31 bits (for slv and unsigned) and 32 bits (for signed) .

File: AlertLogPkg.vhd
Comment (JimLewis) [inline AlertLogPkg.vhd:4618]: As in the previous version, I am not in favor of adding a ";" to the message.    If someone wants a ";" as a separator, they can add one.

I might be agreeable to a "."   However, that would create a problem for others who wanted or already had something else.

**Review by JimLewis — CHANGES_REQUESTED** [review]

File: AlertLogPkg.vhd
Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: I think maybe we need to discuss this.

Actual and Expected are the names used by the scoreboard package.   Many users think in terms of received and expected.   Hence, that is a lesson learned over time and AlertLog came along long after, so it used different naming.

WRT, AlertIf vs AffirmIf.   AffirmIf is intended for self checking.   There is a received/actual and a expected.    

AlertIf started out as general purpose - ie: used for parameter and self-checking, but with the addition of AffirmIf, its intent is parameter checking.   We might be checking that A'length matches B'length.   This is not a received/actual vs expected type of check.

So I think we need to pause and maybe talk about the proposed changes.   GTM?

**Review by JimLewis — CHANGES_REQUESTED** [review]

**Comment by JimLewis** [comment]
@riedel-ferringer  
First, thanks for the pull request.   

Can you submit the changes to ScoreboardGenericPkg and TbUtilPkg separately from AlertLogPkg.   I can merge those.   We need to talk about the changes to AlertLogPkg.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:5927]: Only ModelSim 2021.2, we don't have anything else. The problem was that the resulting string might have wrong range, ie. not starting from (1 to N), but (i+1 to N), which then was a problem at some later point.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:4449]: I just didn't like that the message and "Actual = ..." is not separated by a delimiter. I have no hard feelings about it :-) It looked to me like it was missed.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3687]: Whether 'actual' or 'received' does not matter to me. I was always confused by L/R because I just didn't know which is which. While it doesn't really matter for practical reasons, it's a question of consistency.

However, I understand that there are implications here, probably also for users who rely on specific output strings.

File: ScoreboardGenericPkg.vhd
Comment (riedel-ferringer) [reply ScoreboardGenericPkg.vhd:2086]: I missed the singleton API, because when I started I think there only was the protected type, so I simply didn't realize it.

I have a testbench for a Verification Unit of the Avalon Bus, and I am testing the VU with different parameters (readLatency, readWaitTime, etc). I always send a couple of data and use a scoreboard to check them. When done, I've written a procedure that prints a summary (checked vs failed). I then clear the counters and reuse the scoreboard for the next run with a different avalon-setting.

I think the internal check count is very important, because if we only count error, you wouldn't notice if nothing had been checked at all. I guess the AlertLogPkgs CheckCount is not that well suited for such a check.

File: TbUtilPkg.vhd
Comment (riedel-ferringer) [reply TbUtilPkg.vhd:645]: Already done.

File: TbUtilPkg.vhd
Comment (riedel-ferringer) [reply TbUtilPkg.vhd]: done.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:4618]: good point. I will update this and remove it again.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3822]: replaced everything with to_string for consistency

**Comment by JimLewis** [comment]
I have accepted your changes for TbUtilPkg and ScoreboardGenericPkg.   

I accepted the intent of your change for AlertLogPkg, PathTail.   In its current form it crashed GHDL, so I rewrote it using a variable instead.  

WRT, ScoreboardGenericPkg, at this time, neither the other pull request nor I added SetCheckCountZero to the singleton.   I did make it available to indexed scoreboards that use ScoreboardPType.

WRT, AlertLogPkg, and printing for AlertIfEqual and AlertIfNotEqual and using types that are arrays of std_ulogic/std_logic, I changed the to_string to to_hstring to be consistent.    In the future, maybe we want to add a format specifier?

Currently only ScoreboardGenericPkg uses the word Actual and it only uses it for the type.   When printing, it prints "Received".  So have is currently printed by AffirmIfEqual is consistent with what the scoreboard prints.   If we do a consistency update, it would be to change the Scoreboard parameter to ReceivedType.    Unlikely it will change for ScoreboardPType, however, ...

I have been pondering refactoring a new Singleton Scoreboard that has the subprogram generic AffirmIfEqual instead of Match, to_expected_string and to_actual_string.   That will allow people to control whether Received and Expected are only printed on FAILURE or also during PASSED (requested by the DO-254 folks).   If this happens, having local counts for Error and Passed will not be possible, however, we will be able to fetch them and emulating clearing them (by logging their previous value and subtracting).   The question would be at that time, would we need API commands to get the count for the entire test as well as the count for one segment of the test (after clearing).   This is the reason I did not make those values visible in the Singleton API just yet.    If this one does get created, it would be appropriate to change its ActualType to ReceivedType.

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd]: I removed the semicolon, but kept the space as suggested by Jim

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3687]: After reading your last comment, is it agreed to use "received" / "expected" everywhere? If so, I will update the PR accordingly.

Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: I don't have any compelling reasons to either change or not change AlertIfEqual.   

AlertIfEqual was created before AffirmIfEqual.   It's original intent was self-checking, but that usage was deprecated by AffirmIfEqual.    So AffirmIfEqual evolved to use "received" / "expected" as that is appropriate for self-checking.

What does that mean for AlertIfEqual?   My current thinking is that AlertIfEqual is for broader usage than just "received" / "expected".    I would have to grep through my code to see if I even use it or if I have evolved to only using AffirmIfEqual.   

What is the use model for AlertIfEqual / AlertIfNotEqual?  

```vhdl
function Mux2 (
    Sel : std_logic ; 
    A, B : std_logic_vector 
) return std_logic_vector is
begin
    AlertIfNotEqual(A'length, B'length, "Mux2: Parameter Length Not Matching. ") ; 
```

Here "received" / "expected" is not appropriate.   Hence, maybe it is not so good to change AlertIfEqual/AlertIfNotEqual.

File: ScoreboardGenericPkg.vhd
Comment (JimLewis) [reply ScoreboardGenericPkg.vhd:2086]: One of the next planned additions to reporting is scoreboard reporting.   For the predefined scoreboards this will be automatic.   However, it will only be automatic for the Singleton API.   

One of the things you would be able to do then is for each phase of your test, get a new scoreboardID:
```vhdl
SB := NewID("SB:Phase1") ; 
. . . 
SB := NewID("SB:Phase2") ; 
```
As long as you finish each phase with the scoreboard empty, it is ok to just leave the old one behind.
The reporting will be automatic for each ID in the structure.    

For a limited set of scoreboards, this would be ok.   For a large number of scoreboards, I would need
to check to see what the nominal amount of storage a scoreboard ID uses.   It may be very little - 
ie: just the counters and this would be ok.

Each new scoreboard ID has its own AlertLogID.   Requirements Bins currently setting a Passed Goal.
If this is not allowed for a regular AlertLogID, it needs to be extended to do so.   And then all the 
checking would be automatic.

Comment (riedel-ferringer) [reply ScoreboardGenericPkg.vhd:2086]: > As long as you finish each phase with the scoreboard empty, it is ok to just leave the old one behind.
The reporting will be automatic for each ID in the structure.

Just to clarify things (and I understand this is not the actual topic here): You will automatically create a summary report whenever a new ID is created and the last scoreboard before the newly created one is empty? Or will you always create the report once the scorebord becomes empty? What then, do you deallocate the just-reported scoreboard to free up memory and make the slot available again (which is probably a bad idea because of duplicate IDs)?

File: AlertLogPkg.vhd
Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3687]: Personally, I think that 'actual' is a better name than 'received', because it's - imho - more clear, and more boardly usable. Especially from software tests (eg, Unit Tests in C#), the naming is Actual/Expected. 'Actual' can be used where 'received' would be appropriate, but it can also be used in places where 'received' is not the best choice. As mentioned before, I don't like the L/R notation because you don't know which is which. In theory it would be enough to rename "R" to "expected".

I agree with the mentioned use model for AlertIf..., I am using it in the same way as you suggested: For self-checking or checking invariants that are most of the time fullfilled and should not produce unnecessary log-output.

Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: I don't disagree with your understanding of 'actual'.   However, we need to speak the terms used in the hardware community - which is why we use 'Received' instead.

File: ScoreboardGenericPkg.vhd
Comment (JimLewis) [reply ScoreboardGenericPkg.vhd:2086]: The plan is to do the same thing that is done in the Coverage data structure.    When the test completes and the test calls EndOfTestReports (instead of ReportAlerts).   If you look at ReportPkg.EndOfTestReports for coverage it does.   So if there is coverage, then walk the Coverage singleton, find each coverage model and produce a report for it.   Then there is TCL code that converts the coverage model to HTML that runs when simulate completes.

```vhdl
    if GotCoverage then 
      WriteCovYaml (
        FileName     => "./reports/" & GetAlertLogName & "_cov.yml"
      ) ;
    end if ; 
```

So if you look below the above code, you will see the following commented out.   So what we need is a GotScoreboard (which basically checks to see if there are any scoreboards in the singleton), and a WriteScoreboardYaml, which writes out elements of the Scoreboard that we need for reporting - all the count stuff and such, and then the corresponding TCL code to convert the Yaml to HTML.   It will not be too hard, mostly templating off the code already done in the CoveragePkg.   The only hard part is working out the Yaml format and what to report.   Should not take too long.   Just need to work it into the schedule.
```vhdl
--    if work.ScoreboardPkg_slv.GotScoreboard then 
--       work.ScoreboardPkg_slv.WriteScoreboardYaml ;
--    end if ; 
--
--    if work.ScoreboardPkg_int.GotScoreboard then 
--       work.ScoreboardPkg_int.WriteScoreboardYaml ;
--    end if ; 
```

Comment (JimLewis) [reply ScoreboardGenericPkg.vhd:2086]: Currently we do not delete/deallocate items in the singletons.  That is something I am currently working on and expect to be in one of the next two releases.

File: AlertLogPkg.vhd
Comment (f-markus) [reply AlertLogPkg.vhd:3687]: So how should I change the naming? For Alert... we use "actual/expected", and for Affirm... we use "received/expected"? I think that makes most sense wrt. the mentioned use models.

Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3687]: So how should I change the naming? For Alert... we use "actual/expected", and for Affirm... we use "received/expected"? I think that makes most sense wrt. the mentioned use models.

Comment (Paebbels) [reply AlertLogPkg.vhd:3687]: I'm ok with both variants as long as it's not `R/L`.  
Received is a special case of actual.  
Personally I would prefer `actual/expected` as it's also used in other frameworks or in OSVVM's scoreboard.  

But the final decision is up to @JimLewis :).

Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: @Paebbels 
The intended use model for AffirmIf is to cover all uses cases of self checking.   This is where actual/expected or received/expected make sense.

The intended use model for AlertIfEqual/AlertIfNotEqual is for parameter checking.   It is not clear to me how actual/expected or received/expected make any sense in the above parameter check example above.   Let me be honest, I don't particularly like L and R, however, give me something that works.   You can see that they do not work for my example above right?

Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: For AlertIfEqual we could say, 
Parameters expected to be not equal, Param1 = ...   Param2 = ....

For AlertIfNotEqual we could say, 
Parameters expected to be equal, Param1 = ...   Param2 = ....

Comment (JimLewis) [reply AlertLogPkg.vhd:3687]: From an OSVVM verification suite perspective, the impact of changing AlertIfEqual/AlertIfNotEqual is relatively minor.

OTOH, changing AffirmIfEqual will require significant more work in updating the checking.   So it would be a low priority update relative to other tasks that need to be done.

Comment (riedel-ferringer) [reply AlertLogPkg.vhd:3687]: I agree that actual/expected (or whatever) does not make sense for certain use-models of AlertIfEqual. However, the same would be true for AffirmIfEqual. There might be situation where you don't compare against an expected value, but against another 'actual'. Even if this is not the intended use model, it will be used that we (well, I certainly do). And vice versa, I also use AlertIfEqual to check actuals vs. expected values sometimes. Using param1, param2 does not, imho, solve the problem of being ambigous. I think that more often than not, you compare something against an expected value. And if not, it doesn't matter. I don't think that in your MUX example, using 'actual'/'expected' would make the results less clear.

**Comment by Paebbels** [comment]
@riedel-ferringer Jim and I did some branch cleanups like `master` &rarr; `main`. We also renamed `Dev` &rarr; `dev`. This by accident closed this PR.

@JimLewis can you try with maintainer rights if you can pickup this branch/PR and change the source branch name?

**Comment by Paebbels** [comment]
@riedel-ferringer is it correct that after your last force-push the branch has no diff?

**Comment by riedel-ferringer** [comment]
I am not sure if we yet reached a conclusion here? I can provide an updated version of AlertLogPkg.vhd (as plenty has changed and there are merge conflicts) according to whatever naming scheme you prefer.

The change requests are fixed, at least as far as I can tell.

**Comment by riedel-ferringer** [comment]
> @riedel-ferringer is it correct that after your last force-push the branch has no diff?

Yes it is correct. I accidentially merged into the wrong branch and did not realize that it was the one connected to the PR.

**Comment by JimLewis** [comment]
AlertIfEqual uses L and R because the target of AleritIf is for protocol checkers which only report errors and are not part of self checking.   So while L and R are not that descriptive, using Actual and Expected or Received and Expected is not an exact thing either.

AffirmIfEqual uses received and expected.   At least in English, what is meant by received is better understood than actual.   

It is noted that this is inconsistent with what the Scoreboard uses - actual and expected.    However, I expect the Scoreboard to address this in the future (due to other planned changes).  The current Scoreboard falls short of the communities need in flexibility of the message printing.   I have plans to address this with a future addition to the ScoreboardGenericPkg family.

**Pull request closed without merging**
